src/dnmd/deltas.c

```diff
@@ -1,125 +1,314 @@
+        // Strip this high bit if it is set. We don't need it.
+        mdtable_id_t table_id = ExtractTokenType(tk) & 0x7F;
 
-static bool merge_blob_heap(mdcxt_t* cxt, mdcxt_t* delta)
-{
-    if (delta->blob_heap.size == 0)
-        return true;
+        if (table_id < mdtid_First || table_id >= mdtid_End)
+            return false;
+
+        if (token_map->map_cur_by_table[table_id].count == (uint32_t)(-1))
+        {
+            token_map->map_cur_by_table[table_id].start = map_cur;
+            token_map->map_cur_by_table[table_id].count = 1;
+        }
+        else if (previous_table_id != table_id)
+        {
+            // If the set of remapped tokens for this table has already been started, then the previous token
+            // must be from the same table as the current token (tokens are grouped by table).
+            return false;
+        }
+        else
+        {
+            token_map->map_cur_by_table[table_id].count++;
+        }
+
+        previous_table_id = table_id;
+    }
 
-    mdstream_t* base = &cxt->blob_heap;
-    mdstream_t* blob = &delta->blob_heap;
-    return merge_into_base(cxt, base, blob, 0);
+    return true;
 }
 
-static bool merge_us_heap(mdcxt_t* cxt, mdcxt_t* delta)
+static bool resolve_token(enc_token_map_t* token_map, mdToken referenced_token, mdhandle_t delta_image, mdcursor_t* row_in_delta)
 {
-    if (delta->user_string_heap.size == 0)
-        return true;
+    mdtable_id_t type = ExtractTokenType(referenced_token);
+
+    if (type < mdtid_First || type >= mdtid_End)
+        return false;
+
+    uint32_t rid = RidFromToken(referenced_token);
 
-    mdstream_t* base = &cxt->user_string_heap;
-    mdstream_t* blob = &delta->user_string_heap;
-    return merge_into_base(cxt, base, blob, 0);
+    if (rid == 0)
+        return false;
+
+    // If we don't have any EncMap entries for this table,
+    // then the token in the EncLog is the token we need to look up in the delta image to get the delta info to apply.
+    if (token_map->map_cur_by_table[type].count == (uint32_t)(-1))
+    {
+        if (!md_token_to_cursor(delta_image, referenced_token, row_in_delta))
+            return false;
+    }
+    else
+    {
+        mdcursor_t map_record = token_map->map_cur_by_table[type].start;
+        for (uint32_t i = 0; i < token_map->map_cur_by_table[type].count; md_cursor_next(&map_record), i++)
+        {
+            mdToken mappedToken;
+            if (1 != md_get_column_value_as_constant(map_record, mdtENCMap_Token, 1, &mappedToken))
+                return false;
+            
+            assert((mdtable_id_t)(ExtractTokenType(mappedToken) & 0x7f) == type);
+            if (RidFromToken(mappedToken) == rid)
+            {
+                if (!md_token_to_cursor(delta_image, TokenFromRid(i + 1, CreateTokenType(type)), row_in_delta))
+                    return false;
+                
+                return true;
```

Comment:
```suggestion
                return md_token_to_cursor(delta_image, TokenFromRid(i + 1, CreateTokenType(type)), row_in_delta);
```

Can't we just return?